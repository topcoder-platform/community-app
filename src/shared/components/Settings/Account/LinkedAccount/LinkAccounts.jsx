/**
 * Renders 'Link Your Accounts' section.
 */
/* eslint-disable jsx-a11y/interactive-supports-focus */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import _ from 'lodash';
import React from 'react';
import PT from 'prop-types';
import ImageFilter from 'react-image-filter';
import SocialIcons from './SocialIcons';

import './styles.scss';

function getStyleName(account) {
  let style = 'ext-tile';
  if (account.disabled) {
    style += ' disabled';
  } else {
    style += ' enabled';
  }

  if (account.status === 'linked') {
    style += ' connected';
  } else if (account.status === 'pending') {
    style += ' connecting';
  } else if (account.status === 'unlinked' && !account.disabled) {
    style += ' connect';
  }
  return style;
}

export default class LinkAccounts extends React.Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(e, account) {
    const {
      linkExternalAccount,
      profile,
      tokenV3,
      unlinkExternalAccount,
    } = this.props;
    e.preventDefault();
    if (account.status === 'linked') {
      unlinkExternalAccount(
        profile,
        tokenV3,
        account.providerType,
      );
    } else if (!account.disabled && account.status === 'unlinked') {
      linkExternalAccount(
        profile,
        tokenV3,
        account.providerType,
      );
    }
  }

  render() {
    const {
      allLinks,
      supportedAccounts,
    } = this.props;

    let accounts = _.cloneDeep(supportedAccounts);

    _.remove(accounts, al => al.order < 0);

    _.forEach(accounts, (account) => {
      const linkedAccount = _.find(allLinks, p => p.providerType === account.providerType);

      if (!linkedAccount) {
        account.status = 'unlinked'; /* eslint-disable-line no-param-reassign */
      } else {
        account.status = linkedAccount.status; /* eslint-disable-line no-param-reassign */
      }
    });
    accounts = _.sortBy(accounts, 'order');

    const FILTER_GRAYSCALE = [
      1, 0, 0, 0, 0,
      1, 0, 0, 0, 0,
      1, 0, 0, 0, 0,
      0, 0, 0, 1, 0,
    ];

    const FILTER_NONE = [
      1, 0, 0, 0, 0,
      0, 1, 0, 0, 0,
      0, 0, 1, 0, 0,
      0, 0, 0, 1, 0,
    ];

    return (
      <div styleName="external-links">
        {
          _.map(accounts, account => (
            <div key={account.providerType} role="button" onClick={e => this.handleClick(e, account)} styleName={getStyleName(account)}>
              <div styleName={`external-account-box ${account.colorClass}`}>
                <ImageFilter
                  image={SocialIcons[account.providerType]}
                  filter={account.status === 'linked' ? FILTER_NONE : FILTER_GRAYSCALE}
                />
              </div>
              <div styleName="status">
                { account.displayName }
              </div>
            </div>
          ))
        }
      </div>
    );
  }
}

LinkAccounts.propTypes = {
  tokenV3: PT.string.isRequired,
  profile: PT.shape().isRequired,
  allLinks: PT.arrayOf(PT.shape()).isRequired,
  supportedAccounts: PT.arrayOf(PT.shape()).isRequired,
  linkExternalAccount: PT.func.isRequired,
  unlinkExternalAccount: PT.func.isRequired,
};
