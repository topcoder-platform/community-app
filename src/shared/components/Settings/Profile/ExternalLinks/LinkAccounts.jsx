/**
 * Renders 'Link Your Accounts' section.
 */
/* eslint-disable jsx-a11y/interactive-supports-focus */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import _ from 'lodash';
import React from 'react';
import PT from 'prop-types';

import './styles.scss';

function getStyleName(account) {
  let style = 'ext-tile';
  if (account.disabled) {
    style += ' disabled';
  } else {
    style += ' enabled';
  }

  if (account.status === 'linked') {
    style += ' connected';
  } else if (account.status === 'pending') {
    style += ' connecting';
  } else if (account.status === 'unlinked' && !account.disabled) {
    style += ' connect';
  }
  return style;
}

export default class LinkAccounts extends React.Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(e, account) {
    e.preventDefault();
    if (account.status === 'linked') {
      this.props.unlinkExternalAccount(
        this.props.profile,
        this.props.tokenV3,
        account.providerType,
      );
    } else if (!account.disabled && account.status === 'unlinked') {
      this.props.linkExternalAccount(
        this.props.profile,
        this.props.tokenV3,
        account.providerType,
      );
    }
  }

  render() {
    const {
      allLinks,
      supportedAccounts,
    } = this.props;

    let accounts = _.cloneDeep(supportedAccounts);

    _.remove(accounts, al => al.order < 0);

    _.forEach(accounts, (account) => {
      const linkedAccount = _.find(allLinks, p => p.providerType === account.providerType);

      if (!linkedAccount) {
        account.status = 'unlinked'; /* eslint-disable-line no-param-reassign */
      } else {
        account.status = linkedAccount.status; /* eslint-disable-line no-param-reassign */
      }
    });
    accounts = _.sortBy(accounts, 'order');

    return (
      <div styleName="external-links">
        {
          _.map(accounts, account => (
            <div key={account.providerType} role="button" onClick={e => this.handleClick(e, account)} styleName={getStyleName(account)}>
              <div styleName={`external-account-box ${account.colorClass}`}>
                <i className={`fa ${account.className}`} />
                <div styleName="provider">{account.displayName}</div>
              </div>
              {
                account.status === 'linked' &&
                <div styleName="status already-connected">Connected</div>
              }
              {
                account.status === 'pending' &&
                <div styleName="status already-connected">Connecting</div>
              }
              <div styleName="status disconnect">Disconnect</div>
              {
                account.status === 'unlinked' && !account.disabled &&
                <div styleName="status">Connect</div>
              }
              {
                account.disabled &&
                <div styleName="status">Coming Soon</div>
              }
            </div>
          ))
        }
      </div>
    );
  }
}

LinkAccounts.propTypes = {
  tokenV3: PT.string.isRequired,
  profile: PT.shape().isRequired,
  allLinks: PT.arrayOf(PT.shape()).isRequired,
  supportedAccounts: PT.arrayOf(PT.shape()).isRequired,
  linkExternalAccount: PT.func.isRequired,
  unlinkExternalAccount: PT.func.isRequired,
};
