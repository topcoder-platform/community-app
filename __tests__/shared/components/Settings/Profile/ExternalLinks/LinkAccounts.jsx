import React from 'react';
import Renderer from 'react-test-renderer/shallow';

import LinkAccounts from 'components/Settings/Profile/ExternalLinks/LinkAccounts';

import userProfile from '../../__mocks__/user-profile.json';

const rnd = new Renderer();

it('renders link accounts section of profile setting page correctly', () => {
  rnd.render((<LinkAccounts
    tokenV3=""
    profile={userProfile}
    allLinks={[]}
    supportedAccounts={[]}
    linkExternalAccount={() => {}}
    unlinkExternalAccount={() => {}}
  />));
  expect(rnd.getRenderOutput()).toMatchSnapshot();
});
