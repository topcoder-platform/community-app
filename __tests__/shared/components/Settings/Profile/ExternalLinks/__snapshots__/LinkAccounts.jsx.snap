// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders link accounts section of profile setting page correctly 1`] = `
<div
  className="src-shared-components-Settings-Profile-ExternalLinks-___styles__external-links___1Yaua"
/>
`;
